packages/backend/src/auth/auth.service.ts

```diff
@@ -0,0 +1,152 @@
+import express from 'express'
+import { sign } from 'jsonwebtoken'
+import env from '../config/env'
+import { UnauthorisedException } from '../errors/unauthorisedException'
+import { zParse } from '../middlewares/validator'
+import { User } from '../user/models/user'
+import logger from '../utils/logger'
+import { RefreshToken } from './models/refreshToken'
+import { loginSchema, signupSchema } from './schemas'
+import { BadRequestException } from '../errors/badRequestException'
+
+const log = logger('AuthService')
+
+const generateJWT = (
+  userId: string
+): { accessToken: string; expiresIn: number } => {
+  return {
+    accessToken: sign({ userId }, env.JWT_ACCESS_TOKEN_SECRET, {
+      expiresIn: env.JWT_ACCESS_TOKEN_EXPIRATION_TIME
+    }),
+    expiresIn: Number(env.JWT_ACCESS_TOKEN_EXPIRATION_TIME)
+  }
+}
+
+const generateRefreshToken = (
+  userId: string
+): { refreshToken: string; expiresIn: number } => {
+  return {
+    refreshToken: sign({ userId }, env.JWT_REFRESH_TOKEN_SECRET, {
+      expiresIn: env.JWT_REFRESH_TOKEN_EXPIRATION_TIME
+    }),
+    expiresIn: Number(env.JWT_REFRESH_TOKEN_EXPIRATION_TIME)
+  }
+}
+
+export const signup = async (req: express.Request, res: express.Response) => {
+  try {
+    const { email, password, confirmPassword } = await zParse(signupSchema, req)
+    if (password !== confirmPassword) {
```

Comment:
is it possible to have this check in validation?

Reply:
I tried to do it as Radu suggested, with the Zod object chained with a .superRefine, as in this screenshot.

But this modifies the return type of the Zod object, and it is no longer compatible with the z.infer that we use to keep the types returned by the zParse.

That is why I moved it just as a validation in the controller for now.